### PR TITLE
16-July-2021

### PR DESCRIPTION
**Pull request by PriyangaPKini**
Build a binary search tree
The tree should support the following operations:
- Inserting an integer into the tree
- Finding if a given integer is present in the tree
- Deleting an integer from the tree
The code must have the following characteristics
- The tree must be immutable
- The tree must be thread safe (two threads can modify the tree simultaneously)
- All functions in the code must have a single exit point
- Pay attention to memory usage
Extra credit:
- The tree must be self balancing
- Use your tree to find the count of the words in a huge text file

### CODE REVIEW

**Review by ssrihari — CHANGES_REQUESTED** [review]
reimplement using only what you know, no shortcuts

File: problems/bst/avl-bst.clj [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+;; Extra credit:
+;; - The tree must be self balancing
+;; - Use your tree to find the count of the words in a huge text file
+
+
+(defn insert-node
+  "Returns a bst after inserting a new node"
+  [{:keys [root left right] :as tree} value]
+  (cond
+    (nil? root) (into {} {:root value :left nil :right nil})
```
Comment (ssrihari) [inline problems/bst/avl-bst.clj:20]: you don't need `into` here, you can just return the literal map

File: problems/bst/avl-bst.clj [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    (< value root) (merge tree {:left (insert-node left value)})
```
Comment (ssrihari) [inline problems/bst/avl-bst.clj:21]: look into `assoc`

Comment (ssrihari) [reply problems/bst/avl-bst.clj:21]: and `update`

File: problems/bst/avl-bst.clj [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    (> value root) (merge tree {:right (insert-node right value)})
+    :else tree))
+
+;; test case
```
Comment (ssrihari) [inline problems/bst/avl-bst.clj:25]: like I said earlier, put such code into a `(comment)` block

File: problems/bst/avl-bst.clj [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+(reduce insert-node {} '(5 3 2 4 1))
```
Comment (ssrihari) [inline problems/bst/avl-bst.clj:26]: make a function `create-tree` out of this?

**Review by ssrihari — COMMENTED** [review]
good progress!

File: problems/bst/avl-bst.clj (hunk 1 above)
Comment (PriyangaPKini) [reply problems/bst/avl-bst.clj:20]: okay

File: problems/bst/avl-bst.clj (hunk 2 above)
Comment (PriyangaPKini) [reply problems/bst/avl-bst.clj:21]: ok

File: problems/bst/avl-bst.clj (hunk 4 above)
Comment (PriyangaPKini) [reply problems/bst/avl-bst.clj:26]: sure

File: problems/bst/avl-bst.clj (hunk 3 above)
Comment (PriyangaPKini) [reply problems/bst/avl-bst.clj:25]: ok

File: problems/bst/avl-bst.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (PriyangaPKini) [reply problems/bst/avl-bst.clj:21]: okay

**Pull request merged**